File: 67_AddBinary/addBinary.py
class Solution:
    def addBinary(self, a: str, b: str) -> str:
        i, j = len(a) - 1, len(b) - 1
        res = ''
        carry = 0
        # 从尾到头对位相加
        while i >= 0 or j >= 0:
            n1 = int(a[i]) if i >= 0 else 0
            n2 = int(b[j]) if j >= 0 else 0
            tmp = n1 + n2 + carry
            # 二进制进位
            carry = tmp // 2
            res = str(tmp % 2) + res
            i, j = i - 1, j - 1
        # 最高位如果有进位，需要加上carry
        return res if not carry else '1' + res

    
# 解法二： 转换为数字直接相加
class Solution:
    def addBinary(self, a, b) -> str:
        return '{0:b}'.format(int(a, 2) + int(b, 2))

File: 67_AddBinary/addBinary.go
import "strconv"

// 模拟加法的对位相加 + 进位 的运算过程
func addBinary(a string, b string) string {
    i, j := len(a) - 1, len(b) - 1
    ans := ""

    var a1, b1, sum, carry int
    for i >= 0 || j >= 0 {
        // a1提前结束，则当前位置0
        if i < 0 {
            a1 = 0
        } else {
            a1, _ = strconv.Atoi(string(a[i]))
        }
        // b1提前结束，则当前位置0
        if j < 0 {
            b1 = 0
        } else {
            b1, _ = strconv.Atoi(string(b[j]))
        }

        sum = a1 + b1 + carry
        // 结算下一轮对位相加的进位
        carry = sum / 2
        // 当前位结果 = 总和sum模2
        ans = strconv.Itoa(sum % 2) + ans

        // 下标左移，准备下一轮扫描
        i, j = i - 1, j - 1
    }

    // 上面的结果ans第一位是字符"0"，有进位，则要加入进位"1"
    if carry > 0 {
        ans = "1" + ans
    }

    return ans
}
